src/ILCompiler.Compiler/src/Compiler/ExportedMethodsRootProvider.cs

```diff
@@ -36,11 +45,45 @@ public void AddCompilationRoots(IRootingServiceProvider rootProvider)
                     if (ecmaMethod.IsNativeCallable)
                     {
                         string nativeCallableExportName = ecmaMethod.GetNativeCallableExportName();
+
                         if (nativeCallableExportName != null)
+                        {
+                            _methods.Add(ecmaMethod);
                             rootProvider.AddCompilationRoot(method, "Native callable", nativeCallableExportName);
+                        }
                     }
                 }
             }
+
+            // Only emit exported methods for user supplied assemblies
+            // _exportsFile is null when we're not doing a shared library build
+            if (_module != _module.Context.SystemModule && _exportsFile != null)
```

Comment:
`_module != _module.Context.SystemModule` part of the condition should not be required (once the other part of the feedback is addressed).